vllm/engine/metrics/metrics_registry.py

```diff
@@ -0,0 +1,97 @@
+from aioprometheus import Counter, Gauge, Histogram
+from vllm.engine.metrics.metrics import (CounterMetric, GaugeMetric,
+                                         HistogramMetric)
+
+# The begin-* and end* here are used by the documentation generator
+# to extract the metrics definitions.
+
+# begin-metrics-definitions
+counter_prompt_tokens = Counter("vllm_prompt_tokens_total",
+                                "Number of prefill tokens processed.")
+counter_generation_tokens = Counter("vllm_generation_tokens_total",
+                                    "Number of generation tokens processed.")
+
+gauge_scheduler_running = Gauge(
+    "vllm_requests_running_total",
+    "Number of requests currently running on GPU.")
+gauge_scheduler_swapped = Gauge("vllm_requests_swapped_total",
+                                "Number of requests swapped to CPU.")
+gauge_scheduler_waiting = Gauge("vllm_requests_waiting_total",
+                                "Number of requests waiting to be processed.")
+
+gauge_gpu_cache_usage = Gauge(
+    "vllm_gpu_cache_usage_perc",
+    "GPU KV-cache usage. 1 means 100 percent usage.")
+gauge_cpu_cache_usage = Gauge(
+    "vllm_cpu_cache_usage_perc",
+    "CPU KV-cache usage. 1 means 100 percent usage.")
+
+histogram_time_to_first_token = Histogram(
+    "vllm_time_to_first_token_seconds",
+    "Histogram of time to first token in seconds.",
+    buckets=[0.05, 0.1, 0.25, 0.5, 0.75, 1.0, 2.5, 5.0, 7.5, 10.0])
+histogram_time_per_output_tokens = Histogram(
+    "vllm_time_per_output_tokens_seconds",
+    "Histogram of time per output token in seconds.",
+    buckets=[
+        0.01, 0.025, 0.05, 0.075, 0.1, 0.15, 0.2, 0.3, 0.4, 0.5, 0.75, 1.0, 2.5
+    ])
+histogram_e2e_request_latency = Histogram(
+    "vllm_e2e_request_latency_seconds",
+    "Histogram of end to end request latency in seconds.",
+    buckets=[1.0, 2.5, 5.0, 10.0, 15.0, 20.0, 30.0, 40.0, 50.0, 60.0])
+# end-metrics-definitions
+
+METRICS_REGISTRY = [
+    CounterMetric(
+        counter=counter_prompt_tokens,
+        attr="num_prompt_tokens",
+        template="Prompt throughput: {:0.2f} tok/sec",
+    ),
```

Comment:
Overall I don't think it is a good idea to combine the log output and metrics exported to Prometheus. The log lines are supposed to short enough to fit on one line in the terminal. They show you the server is doing something. There would be a lot more metrics that need to exported to prometheus that should not be in the log line.

Reply:
I agree there are more metrics that should be logged to prometheus than to the log line. `PrometheusMetric` supports an argument called `log_local`. It defaults to True, but if set to False, that metric will not be logged locally.
[You can see it in use for the new metrics I added](https://github.com/vllm-project/vllm/pull/2316/files/14f36aef932105523c837e8c368a7fb82b6cef08#diff-8e77b0181f60542716ba32938f5fa5403b1a88927c3cb5d277133591eee1f5ecR95).  As you can see, none of the new percentile metrics I added are logged to the command line.

I think its a good idea to have each metric be responsible for both logging to prometheus and logging locally. The reason for this is twofold:

- The logic for computing summaries of the individual stats (to be logged locally (e.g. last value for gauges, avg throughput for counters, mean for histograms) is now handled in a generic way as opposed to carrying around lists of previous values in `llm_engine` (like `num_generation_tokens` and `num_prompt_tokens` ... as we add more and more metrics like for all these histograms, we start having to carry around a lot of different attributes in `llm_engine` which can get unweildy

- that it is now very easy to add new metrics to the `metrics_registry` moving forward.